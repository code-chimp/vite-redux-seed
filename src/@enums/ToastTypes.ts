enum ToastTypes {
  Error = 'danger',
  Info = 'info',
  Success = 'success',
  Warning = 'warning',
}

export default ToastTypes;
