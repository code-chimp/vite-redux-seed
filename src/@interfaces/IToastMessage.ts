import ToastType from '../@types/ToastType';

interface IToastMessage {
  id: string;
  type?: ToastType;
  text?: string;
}

export default IToastMessage;
