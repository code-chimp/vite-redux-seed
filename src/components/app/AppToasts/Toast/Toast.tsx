import React, { FC, useEffect, useRef } from 'react';
import { Toast as BSToast } from 'bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faCircleCheck,
  faCircleInfo,
  faCircleXmark,
  faTriangleExclamation,
  IconDefinition,
} from '@fortawesome/free-solid-svg-icons';
import ToastTypes from '../../../../@enums/ToastTypes';
import IToastMessage from '../../../../@interfaces/IToastMessage';
import { useAppDispatch } from '../../../../helpers';
import { removeToastMessage } from '../../../../store/slices/toasts';

export interface IAppToastProps {
  toastMessage: IToastMessage;
}

const Toast: FC<IAppToastProps> = ({ toastMessage }) => {
  const ref = useRef<HTMLDivElement>(null);
  const dispatch = useAppDispatch();
  let icon: IconDefinition;
  let headerText: string;
  let headerClasses = ['toast-header', ' bg-opacity-25'];

  useEffect(() => {
    const handleClose = () => {
      /* third party functionality not executed by Jest */
      /* istanbul ignore next */
      dispatch(removeToastMessage(toastMessage.id));
    };
    const el = ref.current;

    el!.addEventListener('hidden.bs.toast', handleClose);

    const toast = new BSToast(el!);
    toast.show();

    return () => {
      el!.removeEventListener('hidden.bs.toast', handleClose);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  switch (toastMessage.type) {
    case ToastTypes.Error:
      icon = faCircleXmark;
      headerText = 'Error';
      headerClasses = [...headerClasses, 'bg-danger', 'text-danger'];
      break;

    case ToastTypes.Success:
      icon = faCircleCheck;
      headerText = 'Success';
      headerClasses = [...headerClasses, 'bg-success', 'text-success'];
      break;

    case ToastTypes.Warning:
      icon = faTriangleExclamation;
      headerText = 'Warning';
      headerClasses = [...headerClasses, 'bg-warning', 'text-warning'];
      break;

    default:
      icon = faCircleInfo;
      headerText = 'Information';
      headerClasses = [...headerClasses, 'bg-info', 'text-primary'];
  }

  return (
    <div
      ref={ref}
      data-testid={`toast-${toastMessage.id}`}
      className="toast"
      role="alert"
      aria-live="assertive"
      aria-atomic="true">
      <div data-testid={`toast-${toastMessage.id}-header`} className={headerClasses.join(' ')}>
        <FontAwesomeIcon
          data-testid={`toast-${toastMessage.id}-icon`}
          className="flex-shrink-0 me-2"
          icon={icon}
        />
        <strong data-testid={`toast-${toastMessage.id}-header-text`} className="me-auto">
          {headerText}
        </strong>
        <button
          type="button"
          data-testid={`toast-${toastMessage.id}-close`}
          className="btn-close"
          data-bs-dismiss="toast"
          aria-label="Close"></button>
      </div>
      <div data-testid={`toast-${toastMessage.id}-body`} className="toast-body">
        {toastMessage.text}
      </div>
    </div>
  );
};

export default Toast;
