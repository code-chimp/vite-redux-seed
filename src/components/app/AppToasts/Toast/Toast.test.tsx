import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import configureMockStore, { MockStore } from 'redux-mock-store';
import { Provider } from 'react-redux';
import Toast from './Toast';
import IToastMessage from '../../../../@interfaces/IToastMessage';
import ToastTypes from '../../../../@enums/ToastTypes';

const mockStore = configureMockStore([]);
let store: MockStore;

const initialState: Array<IToastMessage> = [
  {
    id: 'foo',
    text: 'you broke something',
    type: ToastTypes.Error,
  },
  {
    id: 'bar',
    text: 'all the things are fragile',
    type: ToastTypes.Info,
  },
  {
    id: 'baz',
    text: 'you stopped breaking things for the day',
    type: ToastTypes.Success,
  },
  {
    id: 'fibbity',
    text: 'you are about to break it',
    type: ToastTypes.Warning,
  },
];

describe('components / app / AppToasts / Toast', () => {
  it('should match the snapshot', () => {
    store = mockStore({ toasts: initialState });
    const { asFragment } = render(
      <Provider store={store}>
        <Toast toastMessage={initialState[0]} />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render provided text in the body', () => {
    store = mockStore({ toasts: initialState });
    const target = initialState[0];

    render(
      <Provider store={store}>
        <Toast toastMessage={target} />
      </Provider>,
    );

    const body = screen.getByTestId(`toast-${target.id}-body`);

    expect(body).not.toBeNull();
    expect(body.textContent).toEqual(target.text);
  });

  it('should render an error toast with the appropriate icon and header', () => {
    store = mockStore({ toasts: initialState });
    const target = initialState[0];

    render(
      <Provider store={store}>
        <Toast toastMessage={target} />
      </Provider>,
    );

    const icon = screen.getByTestId(`toast-${target.id}-icon`);
    const header = screen.getByTestId(`toast-${target.id}-header`);
    const headerText = screen.getByTestId(`toast-${target.id}-header-text`);

    expect(icon).not.toBeNull();
    expect(icon!.classList.contains('fa-circle-xmark')).toBe(true);
    expect(header).not.toBeNull();
    expect(header!.classList.contains(`text-${target.type}`)).toBe(true);
    expect(header!.classList.contains(`bg-${target.type}`)).toBe(true);
    expect(headerText).not.toBeNull();
    expect(headerText.textContent).toEqual('Error');
  });

  it('should render an info toast with the appropriate icon and header', () => {
    store = mockStore({ toasts: initialState });
    const target = initialState[1];

    render(
      <Provider store={store}>
        <Toast toastMessage={target} />
      </Provider>,
    );

    const icon = screen.getByTestId(`toast-${target.id}-icon`);
    const header = screen.getByTestId(`toast-${target.id}-header`);
    const headerText = screen.getByTestId(`toast-${target.id}-header-text`);

    expect(icon).not.toBeNull();
    expect(icon!.classList.contains('fa-circle-info')).toBe(true);
    expect(header).not.toBeNull();
    expect(header!.classList.contains('text-primary')).toBe(true);
    expect(header!.classList.contains(`bg-${target.type}`)).toBe(true);
    expect(headerText).not.toBeNull();
    expect(headerText.textContent).toEqual('Information');
  });

  it('should render a success toast with the appropriate icon and header', () => {
    store = mockStore({ toasts: initialState });
    const target = initialState[2];

    render(
      <Provider store={store}>
        <Toast toastMessage={target} />
      </Provider>,
    );

    const icon = screen.getByTestId(`toast-${target.id}-icon`);
    const header = screen.getByTestId(`toast-${target.id}-header`);
    const headerText = screen.getByTestId(`toast-${target.id}-header-text`);

    expect(icon).not.toBeNull();
    expect(icon!.classList.contains('fa-circle-check')).toBe(true);
    expect(header).not.toBeNull();
    expect(header!.classList.contains(`text-${target.type}`)).toBe(true);
    expect(header!.classList.contains(`bg-${target.type}`)).toBe(true);
    expect(headerText).not.toBeNull();
    expect(headerText.textContent).toEqual('Success');
  });

  it('should render a warning toast with the appropriate icon and header', () => {
    store = mockStore({ toasts: initialState });
    // eslint-disable-next-line no-magic-numbers
    const target = initialState[3];

    render(
      <Provider store={store}>
        <Toast toastMessage={target} />
      </Provider>,
    );

    const icon = screen.getByTestId(`toast-${target.id}-icon`);
    const header = screen.getByTestId(`toast-${target.id}-header`);
    const headerText = screen.getByTestId(`toast-${target.id}-header-text`);

    expect(icon).not.toBeNull();
    expect(icon!.classList.contains('fa-triangle-exclamation')).toBe(true);
    expect(header).not.toBeNull();
    expect(header!.classList.contains(`text-${target.type}`)).toBe(true);
    expect(header!.classList.contains(`bg-${target.type}`)).toBe(true);
    expect(headerText).not.toBeNull();
    expect(headerText.textContent).toEqual('Warning');
  });

  it('should dispatch an action when close button is clicked', () => {
    store = mockStore({ toasts: initialState });
    const target = initialState[0];

    render(
      <Provider store={store}>
        <Toast toastMessage={target} />
      </Provider>,
    );
    // const expected = [
    //   {
    //     type: toasts.actions.removeToastMessage.type,
    //     payload:target.id,
    //   },
    // ];

    const closeButton = screen.getByTestId(`toast-${target.id}-close`);
    expect(closeButton).not.toBeNull();

    fireEvent.click(closeButton!);

    // NOTE: the actual action is not getting dispatched as it depends on the
    //       Bootstrap JavaScript which is not being executed by the test framework
    // expect(store.getActions()).toEqual(expected);
    expect(store.getActions()).toEqual([]);
  });
});
