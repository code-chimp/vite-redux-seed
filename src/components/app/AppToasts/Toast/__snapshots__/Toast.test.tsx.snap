// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components / app / AppToasts / Toast > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    aria-atomic="true"
    aria-live="assertive"
    class="toast fade show showing"
    data-testid="toast-foo"
    role="alert"
  >
    <div
      class="toast-header  bg-opacity-25 bg-danger text-danger"
      data-testid="toast-foo-header"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-circle-xmark flex-shrink-0 me-2"
        data-icon="circle-xmark"
        data-prefix="fas"
        data-testid="toast-foo-icon"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM175 175c9.4-9.4 24.6-9.4 33.9 0l47 47 47-47c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9l-47 47 47 47c9.4 9.4 9.4 24.6 0 33.9s-24.6 9.4-33.9 0l-47-47-47 47c-9.4 9.4-24.6 9.4-33.9 0s-9.4-24.6 0-33.9l47-47-47-47c-9.4-9.4-9.4-24.6 0-33.9z"
          fill="currentColor"
        />
      </svg>
      <strong
        class="me-auto"
        data-testid="toast-foo-header-text"
      >
        Error
      </strong>
      <button
        aria-label="Close"
        class="btn-close"
        data-bs-dismiss="toast"
        data-testid="toast-foo-close"
        type="button"
      />
    </div>
    <div
      class="toast-body"
      data-testid="toast-foo-body"
    >
      you broke something
    </div>
  </div>
</DocumentFragment>
`;
